File: src/dataset/cpp/linked-list_exercism-toemm/gemini/GeminiFlash_linked_list_v2.cpp
#include "linked_list.h"
#include <algorithm> // for std::swap
#include <stdexcept> // for std::out_of_range

namespace linked_list {

    node::node(int data, node* next) : data(data), next(next) {}

    list::list() : head(nullptr), tail(nullptr), count(0) {}

    list::~list() {
        while (head) {
            node* temp = head;
            head = head->next;
            delete temp;
        }
    }

    int list::size() const {
        return count;
    }

    bool list::empty() const {
        return count == 0;
    }

    void list::push_front(int data) {
        node* new_node = new node(data, head);
        head = new_node;
        if (!tail) {
            tail = head;
        }
        ++count;
    }

    void list::push_back(int data) {
        node* new_node = new node(data, nullptr);
        if (tail) {
            tail->next = new_node;
            tail = new_node;
        } else {
            head = tail = new_node;
        }
        ++count;
    }

    int list::front() const {
        if (!head) {
            throw std::out_of_range("front() called on empty list");
        }
        return head->data;
    }

    int list::back() const {
        if (!tail) {
            throw std::out_of_range("back() called on empty list");
        }
        return tail->data;
    }

    void list::pop_front() {
        if (!head) {
            throw std::out_of_range("pop_front() called on empty list");
        }
        node* temp = head;
        head = head->next;
        delete temp;
        --count;
        if (!head) {
            tail = nullptr;
        }
    }

    void list::pop_back() {
        if (!tail) {
            throw std::out_of_range("pop_back() called on empty list");
        }
        if (head == tail) {
            delete head;
            head = tail = nullptr;
            count = 0;
            return;
        }

        node* current = head;
        while (current->next != tail) {
            current = current->next;
        }
        delete tail;
        tail = current;
        tail->next = nullptr;
        --count;
    }
}  // namespace linked_list

File: src/dataset/python/linked-list_exercism-fruit-in/gemini/GeminiFlash_linked_list_v2.py
class Node:
    def __init__(self, value):
        self.val = value
        self.next = None
        self.prev = None


class LinkedList:
    def __init__(self):
        self.front = None
        self.back = None
        self.length = 0

    def push(self, value):
        new_node = Node(value)
        if self.back is None:
            self.front = new_node
            self.back = new_node
        else:
            new_node.prev = self.back
            self.back.next = new_node
            self.back = new_node
        self.length += 1

    def pop(self):
        if self.back is None:
            raise IndexError("pop from an empty LinkedList")

        value = self.back.val
        self.back = self.back.prev

        if self.back is None:
            self.front = None
        else:
            self.back.next = None

        self.length -= 1
        return value

    def shift(self):
        if self.front is None:
            raise IndexError("shift from an empty LinkedList")

        value = self.front.val
        self.front = self.front.next

        if self.front is None:
            self.back = None
        else:
            self.front.prev = None

        self.length -= 1
        return value

    def unshift(self, value):
        new_node = Node(value)
        if self.front is None:
            self.front = new_node
            self.back = new_node
        else:
            new_node.next = self.front
            self.front.prev = new_node
            self.front = new_node
        self.length += 1

    def __len__(self):
        return self.length

    def __iter__(self):
        curr = self.front
        while curr:
            yield curr.val
            curr = curr.next